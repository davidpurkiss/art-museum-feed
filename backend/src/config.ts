import 'dotenv/config';

export interface Config {
  harvardArtMuseum: {
    baseUrl: string;
    apiKey: string;
  };
  enableCsrfPrevention: boolean;
}

const config: Config = {
  harvardArtMuseum: {
    baseUrl:
      process.env.HARVARD_ART_MUSEUM_BASE_URL ||
      'https://api.harvardartmuseums.org',
    apiKey: process.env.HARVARD_ART_MUSEUM_API_KEY,
  },
  enableCsrfPrevention: true,
};

export default config;
