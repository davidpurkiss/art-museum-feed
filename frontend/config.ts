export interface Config {
  backendBaseUrl: string;
}

const config: Config = {
  backendBaseUrl:
    process.env.NEXT_PUBLIC_BACKEND_BASE_URL || 'http://localhost:4000/',
};
export default config;
