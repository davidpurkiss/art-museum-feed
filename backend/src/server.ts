import { ApolloServer } from 'apollo-server';
import typeDefs from './typeDefs';
import resolvers from './resolvers';
import HarvardArtMuseumService from './services/harvardArtMuseum';
import config from './config';

const harvardArtMuseumService = new HarvardArtMuseumService(
  config.harvardArtMuseum.baseUrl,
  config.harvardArtMuseum.apiKey
);

const server = new ApolloServer({
  typeDefs,
  resolvers,
  csrfPrevention: config.enableCsrfPrevention,
  healthCheckPath: '/health',
  async onHealthCheck() {
    await harvardArtMuseumService.healthCheck();
  },
});

export default server;
