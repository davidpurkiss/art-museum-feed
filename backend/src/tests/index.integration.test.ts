import nock from 'nock';
import request from 'supertest';
import server from '../server';
import mockObjects from './fixtures/harvardArtMuseum/objects.json';
import config from '../config';

jest.mock('../config', () => ({
  harvardArtMuseum: {
    baseUrl: 'https://api.test.com',
    apiKey: 'test-api-key',
  },
  enableCsrfPrevention: false,
}));

nock(config.harvardArtMuseum.baseUrl)
  .get('/object')
  .query(true)
  .reply(200, mockObjects);

nock(config.harvardArtMuseum.baseUrl)
  .get('/classification')
  .query(true)
  .reply(200);

nock(config.harvardArtMuseum.baseUrl)
  .get('/classification')
  .query(true)
  .reply(504);

describe('service', () => {
  let url;

  beforeAll(async () => {
    // We pass in the port as 0 to let the server pick its own ephemeral port for testing
    const serverInfo = await server.listen({ port: 0 });
    url = serverInfo.url;
  });

  afterAll(async () => {
    await server?.stop();
  });

  describe('objects', () => {
    it('fetches objects from the harvard art museum api', async () => {
      const res = await server.executeOperation({
        query: `query getBooks($page: Int, $size: Int) { objects(page: $page, size: $size) { title } }`,
        variables: { page: 1, size: 10 },
      });
      expect(res).toMatchSnapshot();
    });
  });

  describe('health check', () => {
    it('returns 200 when healthy', async () => {
      // send our request to the url of the test server
      const response = await request(url).get('health');
      expect(response.ok).toBe(true);
      expect(response.status).toBe(200);
    });

    it('returns 503 when not healthy', async () => {
      // send our request to the url of the test server
      const response = await request(url).get('health');
      expect(response.ok).toBe(false);
      expect(response.status).toBe(503);
    });
  });
});
