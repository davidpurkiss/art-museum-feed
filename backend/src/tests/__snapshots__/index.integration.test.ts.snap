// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`service objects fetches objects from the harvard art museum api 1`] = `
Object {
  "data": Object {
    "objects": Object {
      "info": Object {
        "pages": 64,
      },
      "records": Array [
        Object {
          "century": "20th century",
          "culture": "British",
          "dated": "1923",
          "images": Array [
            Object {
              "baseimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV103452_dynmc",
              "date": "2006-07-13",
            },
          ],
          "people": Array [
            Object {
              "birthplace": "Munich, Germany",
              "culture": "British",
              "displaydate": "1860 - 1942",
              "displayname": "Walter Sickert",
              "gender": "male",
              "role": "Artist",
            },
          ],
          "technique": "Etching",
          "title": "Jack Ashore",
        },
        Object {
          "century": "20th century",
          "culture": "British",
          "dated": "c. 1914",
          "images": Array [
            Object {
              "baseimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV103416_dynmc",
              "date": "2006-07-13",
            },
          ],
          "people": Array [
            Object {
              "birthplace": "Munich, Germany",
              "culture": "British",
              "displaydate": "1860 - 1942",
              "displayname": "Walter Sickert",
              "gender": "male",
              "role": "Artist",
            },
          ],
          "technique": "Etching",
          "title": "The Old Middlesex",
        },
        Object {
          "century": "20th century",
          "culture": "British",
          "dated": "c. 1922",
          "images": Array [
            Object {
              "baseimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV103430_dynmc",
              "date": "2006-07-13",
            },
          ],
          "people": Array [
            Object {
              "birthplace": "Munich, Germany",
              "culture": "British",
              "displaydate": "1860 - 1942",
              "displayname": "Walter Sickert",
              "gender": "male",
              "role": "Artist",
            },
          ],
          "technique": "Etching",
          "title": "In Memoriam T. W. Barrett",
        },
        Object {
          "century": "20th century",
          "culture": "British",
          "dated": "1920",
          "images": Array [
            Object {
              "baseimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV103432_dynmc",
              "date": "2006-07-13",
            },
          ],
          "people": Array [
            Object {
              "birthplace": "Munich, Germany",
              "culture": "British",
              "displaydate": "1860 - 1942",
              "displayname": "Walter Sickert",
              "gender": "male",
              "role": "Artist",
            },
          ],
          "technique": "Etching, engraving and drypoint",
          "title": "La Gaîeté Rochechouart",
        },
        Object {
          "century": "20th century",
          "culture": "American",
          "dated": "c. 1973",
          "images": Array [],
          "people": Array [
            Object {
              "birthplace": "Kaunas, Lithuania",
              "culture": "American",
              "displaydate": "1931 - 1978",
              "displayname": "George Maciunas",
              "gender": "male",
              "role": "Artist",
            },
            Object {
              "birthplace": null,
              "culture": "American",
              "displaydate": null,
              "displayname": "Published by Fluxus",
              "gender": "unidentified",
              "role": "Publisher",
            },
          ],
          "technique": "Photolithograph",
          "title": "Grotesque Face Mask",
        },
        Object {
          "century": "18th century",
          "culture": "Italian",
          "dated": "1775",
          "images": Array [
            Object {
              "baseimageurl": "https://nrs.harvard.edu/urn-3:HUAM:776273",
              "date": "2021-08-09",
            },
            Object {
              "baseimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV172842_dynmc",
              "date": "2008-01-17",
            },
          ],
          "people": Array [
            Object {
              "birthplace": "Cabella Ligure",
              "culture": "Italian",
              "displaydate": "1743 - 1790",
              "displayname": "Giovanni David",
              "gender": "male",
              "role": "Artist",
            },
          ],
          "technique": "Etching and aquatint",
          "title": "The Venetian Lady in Masquerade Dress",
        },
        Object {
          "century": "19th century",
          "culture": "German",
          "dated": "1843",
          "images": Array [
            Object {
              "baseimageurl": "https://nrs.harvard.edu/urn-3:HUAM:76470_dynmc",
              "date": "2004-11-17",
            },
          ],
          "people": Array [
            Object {
              "birthplace": "Breslau, Prussia [now Wroclav, Poland]",
              "culture": "German",
              "displaydate": "1815 - 1905",
              "displayname": "Adolph Menzel",
              "gender": "male",
              "role": "Artist",
            },
          ],
          "technique": "Etching",
          "title": "The Willow Tree",
        },
        Object {
          "century": "20th century",
          "culture": "Mexican",
          "dated": "1945",
          "images": Array [],
          "people": Array [
            Object {
              "birthplace": "Mexico City, Mexico",
              "culture": "Mexican",
              "displaydate": "1902 - 1969",
              "displayname": "Leopoldo Méndez",
              "gender": "male",
              "role": "Artist",
            },
          ],
          "technique": "Wood engraving",
          "title": "That Which Must Not Come; Threat over Mexico",
        },
        Object {
          "century": "19th century",
          "culture": "Swiss",
          "dated": "1801",
          "images": Array [
            Object {
              "baseimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV048581_dynmc",
              "date": "2006-05-10",
            },
          ],
          "people": Array [
            Object {
              "birthplace": null,
              "culture": "Swiss",
              "displaydate": "1764 - 1826",
              "displayname": "Conrad Gessner",
              "gender": "male",
              "role": "Artist",
            },
          ],
          "technique": "Lithograph",
          "title": "Cavalry Charge (first version)",
        },
        Object {
          "century": "17th century",
          "culture": "Flemish",
          "dated": "1630",
          "images": Array [
            Object {
              "baseimageurl": "https://nrs.harvard.edu/urn-3:HUAM:INV165696_dynmc",
              "date": "2007-11-09",
            },
          ],
          "people": Array [
            Object {
              "birthplace": "Antwerp",
              "culture": "Flemish",
              "displaydate": "c. 1596 - 1652-3",
              "displayname": "Christoffel Jegher",
              "gender": "male",
              "role": "Artist",
            },
            Object {
              "birthplace": "Siegen, Westphalia",
              "culture": "Flemish",
              "displaydate": "1577 - 1640",
              "displayname": "After Peter Paul Rubens",
              "gender": "male",
              "role": "Artist after",
            },
          ],
          "technique": "Woodcut",
          "title": "Hercules Slaying Discord",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
